content/js3/blocks/simplifying-element-creation/index.md

```diff
@@ -0,0 +1,166 @@
+    2='Use template tags to simplify element initialisation'
++++
+
+We now have a card showing all of the information for one film.
+
+The code we have is quite repetitive and verbose. It does similar things lots of times.
+
+Let's look at two ways we could simplify this code.
+
+### Refactoring: Extraction a function
+
+One way we can simplify this code is to refactor it.
+
+{{<note type="tip" title="Definition: refactoring">}}
+The process of updating our code quality (without changing the implementation) is called **refactoring**.
+{{</note>}}
+
+We can identify things we're doing several times, and extract a function to do that thing for us.
+
+In this example, we keep doing these three things:
+1. Create a new element (sometimes with a different tag name).
+2. Set that element's text content (always to different values).
+3. Appending that element to some parent element (sometimes a different parent).
+
+We could extract a function which does these three things. The things which are different each time need to be parameteres to the function.
+
+We could write a function like this:
+
+```js
+function createChildElement(parentElement, tagName, textContent) {
+    const element = document.createElement(tagName);
+    element.textContent = textContent;
+    parentElement.append(element);
+    return element;
+}
+```
+
+And then rewrite our code to create the card like this:
+
+```js
+const film = {
+  title: "Killing of Flower Moon",
+  director: "Martin Scorsese",
+  times: ["15:35"],
+  certificate: "15",
+  duration: 112,
+};
+
+function createChildElement(parentElement, tagName, textContent) {
+    const element = document.createElement(tagName);
+    element.textContent = textContent;
+    parentElement.append(element);
+    return element;
+}
+
+const card = document.createElement("section");
+
+createChildElement(card, "h3", film.title);
+
+createChildElement(card, "p", `Director: ${film.director}`);
+
+const times = createChildElement(card, "ul", "");
+
+for (const time of film.times) {
+  createChildElement(times, "li", time);
+}
+
+createChildElement(card, "p", `${film.duration} minutes`);
+
+createChildElement(card, "p", `Certificate: ${film.certificate}`);
+
+document.body.append(card);
+```
+
+This code does exactly the same thing as the code we had before. By introducing a function we have introduced some advantages:
+1. Our code is smaller, which can make it easier to read and understand what it's doing.
+2. If we want to change how we create elements (e.g. if we wanted to also add a class attribute for each element), we only need to write the new code one time, not for every element.
+3. We can see that each of the elements is being created the same way. Before, if we wanted to make sure the director element and duration element were created the same, we would have to compare several lines of code. Because we can see they're calling the same function, we know they're made the same way.
```

Comment:
```suggestion
3. 3. We can see that each element is being created the same way. Before, we would have to compare several lines of code to see this. Because we can see they're calling the same function, we know they're made the same way.
```